Zend/zend_execute.c

```diff
@@ -1812,42 +1812,52 @@ static zend_always_inline void zend_fetch_dimension_address_read(zval *result, z
 			}
 		}
 	} else {
-		if (type != BP_VAR_IS && UNEXPECTED(Z_TYPE_P(container) == IS_UNDEF)) {
-			zval_undefined_cv(EX(opline)->op1.var EXECUTE_DATA_CC);
+		if (type != BP_VAR_IS) {
+			if (UNEXPECTED(Z_TYPE_P(container) == IS_UNDEF)) {
+				zval_undefined_cv(EX(opline)->op1.var EXECUTE_DATA_CC);
+			} else if (!Z_ISERROR_P(container)) {
+				zend_error(E_WARNING, "Variable of type %s does not accept array offsets", zend_get_type_by_const(Z_TYPE_P(container)));
+			}
 		}
+
 		if (/*dim_type == IS_CV &&*/ UNEXPECTED(Z_TYPE_P(dim) == IS_UNDEF)) {
 			zval_undefined_cv(EX(opline)->op2.var EXECUTE_DATA_CC);
 		}
-		ZVAL_NULL(result);
+
+		if (type != BP_VAR_IS && (is_ast || ((EX(opline)+1)->op1_type == IS_VAR && EX(opline)->opcode == (EX(opline)+1)->opcode))) {
```

Comment:
I'm not sure about this condition.
From one hand it's not accurate enough (may fail in case of opcode reordering, or insertion of "debug" opcodes),
and at the same time it looks too complex for tun-time.

May be, we can check the whole condition at compile-time and set opline->execute_data = ZEND_DIM_PROPAGATE_ERROR

Reply:
It does lead itself to problems if one were to reorder, or inject debug codes in-between a single string of fetches.  That said, I'm not exactly certain where to check at compile time.  I did just try to set the extended_value (zend_compile.c:zend_delayed_compile_dim)
```
    opline = zend_delayed_emit_op(result, ZEND_FETCH_DIM_R, &var_node, &dim_node);
    opline->extended_value = ZEND_DIM_PROPAGATE_ERROR;
```
But, this sets that extended value for all fetches, including the last one, which leads to error-values being released, rather than properly set to NULL.